In [1]:
import pandas as pd
import sqlite3 as sql

In [2]:
conn=sql.connect('sql-murder-mystery.db')

In [3]:
conn

In [4]:
cursor = conn.cursor()

In [5]:
# Query the schema to get the list of tables in the database
cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
tables = cursor.fetchall()

# Print the list of tables
print("Tables in the database:")
for table in tables:
    print(table[0])

# Loop through the tables to get information about their columns
for table in tables:
    table_name = table[0]
    print("\nColumns in table:", table_name)
    cursor.execute(f"PRAGMA table_info({table_name})")
    columns_info = cursor.fetchall()
    column_names = [column[1] for column in columns_info]
    print(column_names)

Tables in the database:
crime_scene_report
drivers_license
person
facebook_event_checkin
interview
get_fit_now_member
get_fit_now_check_in
income
solution

Columns in table: crime_scene_report
['date', 'type', 'description', 'city']

Columns in table: drivers_license
['id', 'age', 'height', 'eye_color', 'hair_color', 'gender', 'plate_number', 'car_make', 'car_model']

Columns in table: person
['id', 'name', 'license_id', 'address_number', 'address_street_name', 'ssn']

Columns in table: facebook_event_checkin
['person_id', 'event_id', 'event_name', 'date']

Columns in table: interview
['person_id', 'transcript']

Columns in table: get_fit_now_member
['id', 'person_id', 'name', 'membership_start_date', 'membership_status']

Columns in table: get_fit_now_check_in
['membership_id', 'check_in_date', 'check_in_time', 'check_out_time']

Columns in table: income
['ssn', 'annual_income']

Columns in table: solution
['user', 'value']


In [6]:
# Define a function to display the initial observations of a table
def display_initial_observations(table_name):
    # Write a SQL query to select the first few rows of the specified table
    query = f"SELECT * FROM {table_name} LIMIT 5;"

    # Execute the query and fetch the results
    cursor.execute(query)
    results = cursor.fetchall()

    # Convert the results to a pandas DataFrame for better display
    df = pd.DataFrame(results, columns=[i[0] for i in cursor.description])

    # Display the DataFrame
    print(f"Initial observations of the {table_name} table:")
    display(df)

# Specify the table names for which you want to display initial observations
table_names = ['crime_scene_report', 'drivers_license', 'person', 'facebook_event_checkin', 'interview', 'get_fit_now_member', 'get_fit_now_check_in', 'income', 'solution']

# Display initial observations for each table
for table_name in table_names:
    display_initial_observations(table_name)

Initial observations of the crime_scene_report table:


date     type                                        description  \
0  20180115  robbery  A Man Dressed as Spider-Man Is on a Robbery Spree   
1  20180115   murder                  Life? Dont talk to me about life.   
2  20180115   murder  Mama, I killed a man, put a gun against his he...   
3  20180215   murder                         REDACTED REDACTED REDACTED   
4  20180215   murder  Someone killed the guard! He took an arrow to ...   

       city  
0       NYC  
1    Albany  
2      Reno  
3  SQL City  
4  SQL City

Initial observations of the drivers_license table:


id  age  height eye_color hair_color  gender plate_number  car_make  \
0  100280   72      57     brown        red    male       P24L4U     Acura   
1  100460   63      72     brown      brown  female       XF02T6  Cadillac   
2  101029   62      74     green      green  female       VKY5KR     Scion   
3  101198   43      54     amber      brown  female       Y5NZ08    Nissan   
4  101255   18      79      blue       grey  female       5162Z1     Lexus   

  car_model  
0       MDX  
1       SRX  
2        xB  
3     Rogue  
4        GS

Initial observations of the person table:


id                 name  license_id  address_number address_street_name  \
0  10000   Christoper Peteuil      993845             624        Bankhall Ave   
1  10007  Kourtney Calderwood      861794            2791       Gustavus Blvd   
2  10010            Muoi Cary      385336             741     Northwestern Dr   
3  10016          Era Moselle      431897            1987       Wood Glade St   
4  10025         Trena Hornby      550890             276       Daws Hill Way   

         ssn  
0  747714076  
1  477972044  
2  828638512  
3  614621061  
4  223877684

Initial observations of the facebook_event_checkin table:


person_id  event_id                                         event_name  \
0      28508      5880    Nudists are people who wear one-button suits.\n   
1      63713      3865  but that's because it's the best book on anyth...   
2      63713      3999           If Murphy's Law can go wrong, it will.\n   
3      63713      6436  Old programmers never die.  They just branch t...   
4      82998      4470               Help a swallow land at Capistrano.\n   

       date  
0  20170913  
1  20171009  
2  20170502  
3  20170926  
4  20171022

Initial observations of the interview table:


person_id                                         transcript
0      28508                ‘I deny it!’ said the March Hare.\n
1      63713                                                 \n
2      86208      way, and the whole party swam to the shore.\n
3      35267  lessons in here? Why, there’s hardly room for ...
4      33856                                                 \n

Initial observations of the get_fit_now_member table:


id  person_id                name  membership_start_date  \
0  NL318      65076     Everette Koepke               20170926   
1  AOE21      39426        Noe Locascio               20171005   
2  2PN28      63823  Jeromy Heitschmidt               20180215   
3  0YJ24      80651      Waneta Wellard               20171206   
4  3A08L      32858        Mei Bianchin               20170401   

  membership_status  
0              gold  
1           regular  
2            silver  
3              gold  
4            silver

Initial observations of the get_fit_now_check_in table:


membership_id  check_in_date  check_in_time  check_out_time
0         NL318       20180212            329             365
1         NL318       20170811            469             920
2         NL318       20180429            506             554
3         NL318       20180128            124             759
4         NL318       20171027            418            1019

Initial observations of the income table:


ssn  annual_income
0  100009868          52200
1  100169584          64500
2  100300433          74400
3  100355733          35900
4  100366269          73000

Initial observations of the solution table:


Empty DataFrame
Columns: [user, value]
Index: []

In [7]:
# 1. Retrieve Crime Scene Report
crime_scene_query = """
SELECT *
FROM crime_scene_report
WHERE date = '20180115' AND city = 'SQL City'
"""
cursor.execute(crime_scene_query)
crime_scene_report = cursor.fetchall()
print("Crime Scene Report:")
print(crime_scene_report)

Crime Scene Report:
[(20180115, 'assault', 'Hamilton: Lee, do you yield? Burr: You shot him in the side! Yes he yields!', 'SQL City'), (20180115, 'assault', 'Report Not Found', 'SQL City'), (20180115, 'murder', 'Security footage shows that there were 2 witnesses. The first witness lives at the last house on "Northwestern Dr". The second witness, named Annabel, lives somewhere on "Franklin Ave".', 'SQL City')]


Based on this information, we know that there was a murder case on January 15, 2018, in SQL City. The report mentions two witnesses: one residing on "Northwestern Dr" and the other named Annabel living on "Franklin Ave."

In [8]:
# Extracting details of the two witnesses based on their IDs obtained from the previous outputs
witness_ids = [16371, 67318]  # IDs of Annabel Miller and Jeremy Bowers respectively

# Retrieve details of the witnesses
cursor.execute(f"""
    SELECT p.name, p.address_number, p.address_street_name, d.plate_number, d.car_make, d.car_model
    FROM person p
    INNER JOIN drivers_license d ON p.license_id = d.id
    WHERE p.id IN ({', '.join(map(str, witness_ids))})
""")
witness_details = cursor.fetchall()

print("Witness Personal Details:")
for details in witness_details:
    print(details)


Witness Personal Details:
('Annabel Miller', 103, 'Franklin Ave', '23AM98', 'Toyota', 'Yaris')
('Jeremy Bowers', 530, 'Washington Pl, Apt 3A', '0H42W2', 'Chevrolet', 'Spark LS')


Based on the witness personal details provided, Annabel Miller resides at 103 Franklin Ave and drives a Toyota Yaris with license plate 23AM98, while Jeremy Bowers resides at 530 Washington Pl, Apt 3A and drives a Chevrolet Spark LS with license plate 0H42W2.

In [9]:
# IDs of the witnesses based on the provided output
witness_ids = [16371, 67318]  # IDs of Annabel Miller and Jeremy Bowers respectively

# Retrieve details of the witnesses
for idx, witness_id in enumerate(witness_ids, 1):
    cursor.execute(f"""
        SELECT transcript
        FROM interview
        WHERE person_id = {witness_id};
    """)
    witness_interview = cursor.fetchone()

    # Display the transcript of the witness interview
    print(f"\n-- Witness {idx}")
    print(witness_interview[0])


-- Witness 1
I saw the murder happen, and I recognized the killer from my gym when I was working out last week on January the 9th.

-- Witness 2
I was hired by a woman with a lot of money. I don't know her name but I know she's around 5'5" (65") or 5'7" (67"). She has red hair and she drives a Tesla Model S. I know that she attended the SQL Symphony Concert 3 times in December 2017.



Witness 1 claims to have witnessed the murder and recognized the killer from their gym on January 9th, indicating potential gym membership records as a clue.

Witness 2 provides a description of a woman hired for the crime, mentioning her height, red hair, and association with a Tesla Model S, as well as her attendance at the SQL Symphony Concert in December 2017. This information could lead to identifying the suspect through related records such as gym memberships, concert attendance logs, or driver's license details.

In [10]:
# Names of the witnesses based on the provided output
witness_names = ['Annabel Miller', 'Jeremy Bowers']

# Retrieve gym members based on the names of the witnesses
for idx, witness_name in enumerate(witness_names, 1):
    cursor.execute(f"""
        SELECT m.name AS member_name, m.membership_start_date, m.membership_status
        FROM get_fit_now_member AS m
        INNER JOIN person AS p ON m.person_id = p.id
        WHERE p.name = '{witness_name}';
    """)
    gym_members = cursor.fetchall()

    # Display gym members for the witness
    print(f"\n-- Witness {idx} ({witness_name}) Gym Members:")
    for member in gym_members:
        print(member)


-- Witness 1 (Annabel Miller) Gym Members:
('Annabel Miller', 20160208, 'gold')

-- Witness 2 (Jeremy Bowers) Gym Members:
('Jeremy Bowers', 20160101, 'gold')


Annabel Miller and Jeremy Bowers, both gold members, were mentioned in the witness interviews.
This indicates a potential connection between the gym members and the murder case, especially considering their gold membership status. Further investigation is required to determine their involvement.

In [11]:
# Define the check-in date and membership ID pattern
check_in_date = '20180109'
membership_id_pattern = '48Z%'

# Query to check the gym database
gym_check_in_query = f"""
SELECT *
FROM get_fit_now_check_in
WHERE check_in_date = '{check_in_date}'
AND membership_id LIKE '{membership_id_pattern}';
"""

# Execute the query
cursor.execute(gym_check_in_query)
gym_check_ins = cursor.fetchall()

# Display the gym check-in records
print("\nGym Check-in Records:")
for check_in_record in gym_check_ins:
    print(check_in_record)



Gym Check-in Records:
('48Z7A', 20180109, 1600, 1730)
('48Z55', 20180109, 1530, 1700)


The output indicates gym check-in records for the specified date ('20180109') and membership ID pattern ('48Z%'). Two check-in records were found:

Membership ID: '48Z7A', Check-in Date: 20180109, Check-in Time: 16:00, Check-out Time: 17:30

Membership ID: '48Z55', Check-in Date: 20180109, Check-in Time: 15:30, Check-out Time: 17:00

In [12]:
# Define the plate number pattern
plate_number_pattern = '%H42W%'

# Query to check car details
car_details_query = f"""
SELECT *
FROM drivers_license
WHERE plate_number LIKE '{plate_number_pattern}';
"""

# Execute the query
cursor.execute(car_details_query)
car_details = cursor.fetchall()

# Display the car details
print("\nCar Details:")
for car_detail in car_details:
    print(car_detail)



Car Details:
(183779, 21, 65, 'blue', 'blonde', 'female', 'H42W0X', 'Toyota', 'Prius')
(423327, 30, 70, 'brown', 'brown', 'male', '0H42W2', 'Chevrolet', 'Spark LS')
(664760, 21, 71, 'black', 'black', 'male', '4H42WR', 'Nissan', 'Altima')


The output displays car details matching the specified plate number pattern ('%H42W%'). Three records were found:

Plate Number: 'H42W0X', Car Make: Toyota, Car Model: Prius

Plate Number: '0H42W2', Car Make: Chevrolet, Car Model: Spark LS

Plate Number: '4H42WR', Car Make: Nissan, Car Model: Altima

These records provide information about the cars with plate numbers that match the specified pattern.

In [13]:
# Define the license IDs
license_id_male_1 = '423327'
license_id_male_2 = '664760'

# Query to retrieve personal details for Male 1
personal_details_male_1_query = f"""
SELECT *
FROM person
WHERE license_id = '{license_id_male_1}';
"""

# Execute the query for Male 1
cursor.execute(personal_details_male_1_query)
personal_details_male_1 = cursor.fetchall()

# Display the personal details for Male 1
print("\nPersonal Details for Male 1:")
for detail in personal_details_male_1:
    print(detail)

# Query to retrieve personal details for Male 2
personal_details_male_2_query = f"""
SELECT *
FROM person
WHERE license_id = '{license_id_male_2}';
"""

# Execute the query for Male 2
cursor.execute(personal_details_male_2_query)
personal_details_male_2 = cursor.fetchall()

# Display the personal details for Male 2
print("\nPersonal Details for Male 2:")
for detail in personal_details_male_2:
    print(detail)



Personal Details for Male 1:
(67318, 'Jeremy Bowers', 423327, 530, 'Washington Pl, Apt 3A', 871539279)

Personal Details for Male 2:
(51739, 'Tushar Chandra', 664760, 312, 'Phi St', 137882671)


In [14]:
# Define the membership status and ID pattern
membership_status = 'gold'
id_pattern = '48Z%'

# Query to retrieve membership status at the gym
membership_status_query = f"""
SELECT *
FROM get_fit_now_member
WHERE membership_status = '{membership_status}'
AND id LIKE '{id_pattern}';
"""

# Execute the query
cursor.execute(membership_status_query)
membership_status_result = cursor.fetchall()

# Display the membership status at the gym
print("\nMembership Status at the Gym:")
for result in membership_status_result:
    print(result)



Membership Status at the Gym:
('48Z7A', 28819, 'Joe Germuska', 20160305, 'gold')
('48Z55', 67318, 'Jeremy Bowers', 20160101, 'gold')


Based on the information gathered from the investigation:

Witness Testimony:

Witness 1 claims to have witnessed the murder and recognized the killer from their gym on January 9th.
Witness 2 provided a description of a woman hired for the crime, mentioning her height, red hair, and association with a Tesla Model S, as well as her attendance at the SQL Symphony Concert in December 2017.

Gym Records:

Two individuals, Joe Germuska and Jeremy Bowers, have gold membership status at the gym.
Jeremy Bowers, identified as Witness 1, has been mentioned in the witness testimony as recognizing the killer from the gym.

Driver's License Records:
Jeremy Bowers (License ID: 423327) drives a Chevrolet Spark LS with license plate '0H42W2'.
Based on the above information, **Jeremy Bowers**  is likely to be the murderer .